src/providers/ModalProvider/index.tsx

```diff
@@ -0,0 +1,62 @@
+import { createContext, ReactElement, useState, useContext, useCallback } from 'react';
+import Modal from './Modal';
+
+export interface IModalsContext {
+  showModal: (modalContent: ReactElement<any>) => number;
+  closeModal: (id: number) => void;
+  deleteModal: (id: number) => void;
+}
+
+const ModalContext = createContext({} as IModalsContext);
+
+const ModalProvider = ({ children }: { children: JSX.Element }) => {
+  const [modals, setModals] = useState<{ [id: number]: { content: ReactElement<any>, show: boolean, onClose?: () => void } }>({});
+
+  const showModal = useCallback((modalContent: ReactElement<any>, onClose?: () => void) => {
+    const id = new Date().getTime();
+
+    setModals(modals => ({
+      ...modals,
+      [id]: { content: modalContent, show: true, onClose: onClose }
+    }));
+
+    return id;
+  }, [setModals]);
+
+  const closeModal = useCallback((id: number) => {
+    setModals(modals => {
+      const modal = modals[id];
+
+      return {
+        ...modals,
+        [id]: { ...modal, show: false }
+      };
+    });
+  }, [setModals]);
+
+  const deleteModal = useCallback((id: number) => {
+    setModals(modals => {
+      if (modals[id]) {
+        modals[id].onClose && modals[id].onClose!();
```

Comment:
What does ! means?

Reply:
It means that value exists and not null. Without that symbol TS compiler show error.